### PR TITLE
Salesforce - billing information and date joined issue (part 1)

### PR DESCRIPTION
**Pull request by svogt0511**
## Purpose

closes: none

Don't write 'date joined' or 'billing information' to Salesforce from Fabrica if either is blank.  Need to test this before I add it to production.  So far, modification is only in stage.

## Approach

Check quantities in the salesforce api lambda function and write only when conditions are met.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)

- [ ] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Reviewer, please remember our [guidelines](https://datacite.atlassian.net/wiki/spaces/TEC/pages/1168375809/Pull+Request+Guidelines):

- Be humble in the language and feedback you give, ask don't tell.
- Consider using positive language as opposed to neutral when offering feedback. This is to avoid the negative bias that can occur with neutral language appearing negative.
- Offer suggestions on how to improve code e.g. simplification or expanding clarity.
- Ensure you give reasons for the changes you are proposing.

### CODE REVIEW

File: stage/services/salesforce-api/salesforce-api_runner.js [hunk 1]
```diff
@@ -1,3 +1,25 @@
...
+    )
+  ) ? true : false;
+}
+
+function hasjoinedDate(res = new Object({})) {
+  return (
+    res && res.attributes && (
+      res.attributes.joined
+    )
+  ) ? true : false;
```
Comment (mfenner) [inline stage/services/salesforce-api/salesforce-api_runner.js:20]: I don't think this is necessary.

File: stage/services/salesforce-api/salesforce-api_runner.js [hunk 2]
```diff
@@ -1,3 +1,25 @@
+function hasBillingInfo(res = new Object({})) {
+  return (
+    res && res.attributes && (
+      res.attributes.billing_organization ||
+      res.attributes.billing_department ||
+      res.attributes.billing_street ||
+      res.attributes.billing_city ||
+      res.attributes.billing_state_code ||
```
Comment (mfenner) [inline stage/services/salesforce-api/salesforce-api_runner.js:8]: The state is not always required. Only for US, China and some other countries, see https://github.com/datacite/bracco/blob/cd6e57c731ed1699c3368802b3b9a9514a6df470/app/controllers/providers/new.js#L9

**Comment by mfenner** [comment]
Looks good. Consider that state information is only required for some countries.

File: stage/services/salesforce-api/salesforce-api_runner.js (hunk 1 above)
Comment (richardhallett) [reply stage/services/salesforce-api/salesforce-api_runner.js:20]: Ye the extra function isn't needed here, you can just check res.attributes.joined directly.

File: stage/services/salesforce-api/salesforce-api_runner.js [hunk 3]
```diff
@@ -160,9 +186,9 @@ exports.handler = async function (event, context) {
 
     // some member types support billing information
     if (
-      ["Direct Member", "Consortium Organization", "Member Only"].includes(
-        res.attributes.member_type
-      )
+      ["Direct Member", "Consortium Organization", "Member Only"].includes(res.attributes.member_type) &&
+      hasBillingInfo(res) &&
+      hasJoinedDate(res)
```
Comment (richardhallett) [inline stage/services/salesforce-api/salesforce-api_runner.js:191]: So what's the reasoning we dont want to send billing information if there is no date joined?

File: stage/services/salesforce-api/salesforce-api_runner.js [hunk 4]
```diff
@@ -160,9 +186,9 @@ exports.handler = async function (event, context) {
...
     ) {
       body = Object.assign(body, {
         Billing_Organization__c: res.attributes.billing_organization,
```
Comment (richardhallett) [inline stage/services/salesforce-api/salesforce-api_runner.js:194]: Depending on the salesforce API could we just not check individual data like:
Billing_Organization__c: res.attributes.billing_organization ? res.attributes.billing_organization : null

File: stage/services/salesforce-api/salesforce-api_runner.js (hunk 3 above)
Comment (svogt0511) [reply stage/services/salesforce-api/salesforce-api_runner.js:191]: We only want to send billing info if:

- There is a date joined.
- There is billing info in the update.

File: stage/services/salesforce-api/salesforce-api_runner.js (hunk 4 above)
Comment (svogt0511) [reply stage/services/salesforce-api/salesforce-api_runner.js:194]: I am going to leave this the way it is for now.  I think checking individually will still allow billing info in salesforce to be nulled out if all fields are null.

**Pull request merged**